File: PYTHON/169_MajorityElement.py
# Name: Deniz Karakay
# 05/10/2022
# https://leetcode.com/problems/majority-element/

from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        length = len(nums)

        if length == 1:
            return nums[0]

        num_dict = {}

        for num in nums:
            if num in num_dict:
                num_dict[num] += 1
                if num_dict[num] > length // 2:
                    return num
            else:
                num_dict[num] = 1

    def majorityElementAlternative(self, nums: List[int]) -> int:
        num_dict = {}

        for num in nums:
            num_dict[num] = (num_dict.get(num) or 0) + 1

        for n in num_dict:
            if num_dict[n] > len(nums) / 2:
                return n


s = Solution()
print(s.majorityElementAlternative([2, 2, 1, 1, 1, 2, 2]))

File: C++/169_MajorityElement.cpp
/*

Given an array nums of size n, return the majority element.

The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.

Example 1:

Input: nums = [3,2,3]
Output: 3
Example 2:

Input: nums = [2,2,1,1,1,2,2]
Output: 2

Constraints:

n == nums.length
1 <= n <= 5 * 104
-109 <= nums[i] <= 109

Follow-up: Could you solve the problem in linear time and in O(1) space?

*/

class Solution {
public:
    int majorityElement(vector<int>& nums) {
       int n=nums.size();
        int count=0;
        int max=INT_MIN;
        for(int i=0;i<n;i++){
            if (count == 0) {
                max = nums[i];
                count = 1;
                continue;
            }
            if (max != nums[i]) {
                count--;
            } else {
                count++;
            }
        }
        return max;
    }
};